usr/man/man4/ra.4

.\" Copyright (c) 1980 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\"	@(#)ra.4	6.2 (Berkeley) 9/6/87
.\"
.TH RA 4 "September 6, 1987"
.UC 4
.SH NAME
ra \- MSCP disk controller interface
.SH SYNOPSIS
.ft B
.nf
.ta .5i +\w'NRAC 'u +\w'ra_controllers 'u
/sys/conf/SYSTEM:
	NRAC	\fIra_controllers\fP	# NRAD controllers
	NRAD	\fIra_drives\fP	# RX50, RC25, RD51/52/53, RA60/80/81
.DT

/etc/dtab:
.ta .5i +\w'#Name 'u +\w'Unit# 'u +\w'177777 'u +\w'Vector 'u +\w'Br 'u +\w'xxxxxxx 'u +\w'xxxxxxx 'u
	#Name	Unit#	Addr	Vector	Br	Handler(s)		# Comments
	ra	?	172150	154	5	raintr		# uda50, rqdx1/2/3
.DT

major device number(s):
	raw: 14
	block: 5
minor device encoding:
	bits 0007 specify partition of RA drive
	bits 0070 specify RA drive
	bits 0300 specify RA controller
.fi
.ft R
.SH DESCRIPTION
This is a driver for the DEC UDA-50 disk controller
and for other compatible controllers.
The UDA-50 communicates with the host through a packet
oriented protocol termed the Mass Storage Control Protocol (MSCP).
Consult the file
.RI < pdp/mscp.h >
for a detailed description of this protocol.
.PP
Files with minor device numbers 0 through 7 refer to various portions
of drive 0;
minor devices 8 through 15 refer to drive 1, etc.
The standard device names begin with ``ra'' followed by
the drive number and then a letter a-h for partitions 0-7 respectively.
The character ? stands here for a drive number in the range 0-7.
.PP
The block files access the disk via the system's normal
buffering mechanism and may be read and written without regard to
physical disk records.  There is also a `raw' interface
which provides for direct transmission between the disk
and the user's read or write buffer.
A single read or write call results in exactly one I/O operation
and therefore raw I/O is considerably more efficient when
many words are transmitted.  The names of the raw files
conventionally begin with an extra `r.'
.PP
In raw I/O the buffer must begin on a word (even) boundary,
and counts should be a multiple of 512 bytes (a disk sector).
Likewise
.I seek
calls should specify a multiple of 512 bytes.
.SH "DISK SUPPORT"
This driver configures the drive type of each drive
when it is first opened.
A partition table in the driver is required for each type of disk.
The origin and size (in sectors) of the pseudo-disks
on each drive are shown below.
Not all partitions begin on
cylinder boundaries, as on other drives, because previous drivers
used one partition table for all drive types.
Variants of the partition tables are common;
check the driver and the file
.I /etc/disktab
.RI ( disktab (5))
for other possibilities.
.nf
.ta .5i +\w'disk  'u +\w'000000  'u +\w'000000  'u +\w'000000  'u
.PP
.ne 10
RC25 partitions
	disk	 start	   end	length	comments
	ra?a	     0	 15883	 15884	/
	ra?b	 15884	 25915	 10032	swap
	ra?c	 25916	 50901	 24986	/usr
	ra?d	unused
	ra?e	unused
	ra?f	unused
	ra?g	unused
	ra?h	     0	 50901	 50902	whole disk
.PP
.ne 10
RX50 partitions
	disk	 start	   end	length	comments
	ra?a	unusable
	ra?b	unusable
	ra?c	unusable
	ra?d	unusable
	ra?e	unusable
	ra?f	unusable
	ra?g	unusable
	ra?h	     0	  1637	  1638	whole disk
.PP
.ne 10
RD51 partitions
	disk	 start	   end	length	comments
	ra?a	     0	  9699	  9700
	ra?b	 unusable
	ra?c	 unusable
	ra?d	 unusable
	ra?e	     0	  7459	  7460	/
	ra?f	  7460	  9699	  2240	swap
	ra?g	  9700	 21567	 11868	/usr
	ra?h	     0	 21567	 21568	whole disk
.PP
.ne 10
RD52 partitions
	disk	 start	   end	length	comments
	ra?a	     0	  9699	  9700	/
	ra?b	  9700	 26999	 17300
	ra?c	 27000	 30099	  3100	swap
	ra?d	 30100	 60447	 30348	/usr
	ra?e	     0	  7459	  7460
	ra?f	  7460	  9699	  2240
	ra?g	  9700	 60447	 50748
	ra?h	     0	 60447	 60448
.PP
.ne 10
RD53 partitions
	disk	 start	   end	length	comments
	ra?a	     0	  9699	  9700	/
	ra?b	  9700	 26999	 17300
	ra?c	 27000	 30099	  3100	swap
	ra?d	 30100	138639	108540	/usr
	ra?e	     0	  7459	  7460
	ra?f	  7460	  9699	  2240
	ra?g	  9700	138639	128940
	ra?h	     0	138639	138640	whole disk
.PP
.ne 10
RA60 partitions
	disk	 start	   end	length	comments
	ra?a	     0	 15883	 15884	/
	ra?b	 15884	 49323	 33440	swap
	ra?c	 49324	242605	193282	/usr
	ra?d	242606	258489	 15884
	ra?e	258490	399175	140686
	ra?f	unused
	ra?g	242606	399175	156570
	ra?h	     0	399175	399176	whole disk
.PP
.ne 10
RA80 partitions
	disk	 start	   end	length	comments
	ra?a	     0	 15883	 15884	/
	ra?b	 15884	 49323	 33440	swap
	ra?c	 49324	236211	186888	/usr
	ra?d	unused
	ra?e	unused
	ra?f	unused
	ra?g	unused
	ra?h	0	236211	236212	whole disk
.PP
.ne 10
RA81 partitions
	disk	 start	   end	length	comments
	ra?a	     0	 15883	 15884	/
	ra?b	 15884	 49323	 33440	swap
	ra?c	 49324	242605	193282	/usr
	ra?d	242606	258489	 15884
	ra?e	258490	565689	307200
	ra?f	565690	890071	324382
	ra?g	242606	890071	647466
	ra?h	     0	890071	890072	whole disk
.DT
.fi
.PP
The ra?a partition is normally used for the root file system,
the ra?b partition as a paging area,
and the ra?c partition for pack-pack copying (it maps the entire disk).
.SH FILES
.ta \w'/dev/MAKEDEV.local  'u
/dev/ra[0-9][a-f]
.br
/dev/rra[0-9][a-f]
.br
/dev/MAKEDEV	script to create special files
.br
.SH SEE ALSO
hk(4),
ram(4),
rk(4),
rl(4),
rx(4),
si(4),
xp(4),
dtab(5),
autoconfig(8)
.SH DIAGNOSTICS
.BR "rasa %o, state %d" .
(Additional status information given after a hard i/o error.)
The values of the UDA-50 status register and the internal
driver state are printed.
.PP
.BR "ra%d: interrupt in unknown state %d ignored" .
An interrupt was received when the driver was in an unknown
internal state.  Indicates a hardware problem or a driver bug.
.PP
.BR "ra%d: fatal error (%o)" .
The UDA-50 indicated a ``fatal error'' in the status returned
to the host.  The contents of the status register are displayed.
.PP
.BR "ra%d,%d: OFFLINE" .
(Additional status information given after a hard i/o error.)
A hard i/o error occurred because the drive was not on-line.
The attached unit number and the MSCP unit numbers are printed.
.PP
.BR "status %o" .
(Additional status information given after a hard i/o error.)
The status information returned from the UDA-50 is tacked onto
the end of the hard error message printed on the console.
.PP
.BR "ra: unknown packet opcode=0%o" .
An MSCP packet of unknown type was received from the UDA-50.
Check the cabling to the controller.
.PP
The following errors are interpretations of MSCP error messages
returned by the UDA-50 to the host.
.PP
.BR "ra: %s error, controller error, event 0%o" .
.PP
.BR "ra: %s error, host memory access error, event 0%o, addr 0%o" .
.PP
.BR "ra: %s error, disk transfer error, unit %d, grp 0x%x, hdr 0x%x" .
.PP
.BR "ra: %s error, SDI error, unit %d, event 0%o, hdr 0x%x" .
.PP
.BR "ra: %s error, small disk error, unit %d, event 0%o, cyl %d" .
.PP
.BR "ra: %s error, unknown error, unit %d, format 0%o, event 0%o" .
.SH BUGS
In raw I/O
.I read
and
.IR write (2)
truncate file offsets to 512-byte block boundaries,
and
.I write
scribbles on the tail of incomplete blocks.
Thus,
in programs that are likely to access raw devices,
.I read, write
and
.IR lseek (2)
should always deal in 512-byte multiples.
.PP
The partition tables attempt to combine compatibility
with previous drivers and functionality; this is impossible.
The best solution would be to read the partition tables
off the drive.


usr/src/sys/pdpstand/ra.c
/*
 * Copyright (c) 1986 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 *
 *	@(#)ra.c	1.1 (2.10BSD Berkeley) 12/1/86
 */

/*
 * RAxx disk device driver
 * RQDX1 (rx50, rd50, rd51, rd52, rd53)
 */
#include "../h/param.h"
#include "../h/inode.h"
#include "../machine/mscp.h"
#include "../pdpuba/rareg.h"
#include "saio.h"

/*
 * RA Communications Area
*/
struct  rdca {
	short	ca_xxx1;	/* unused */
	char	ca_xxx2;	/* unused */
	char	ca_bdp;		/* BDP to purge */
	short	ca_cmdint;	/* command queue transition interrupt flag */
	short	ca_rspint;	/* response queue transition interrupt flag */
	short	ca_rspl;	/* response descriptors */
	short	ca_rsph;
	short	ca_cmdl;	/* command descriptors */
	short	ca_cmdh;
};

#define	ca_ringbase	ca_rspl

#define	RA_OWN	0x8000	/* RQ owns this descriptor */
#define	RA_INT	0x4000	/* allow interrupt on ring transition */

static	struct	radevice *RDADDR = ((struct radevice *)0172150);

static	struct	ra {
	struct	rdca	ra_ca;
	struct	mscp	ra_rsp;
	struct	mscp	ra_cmd;
} rd;

struct 	mscp 	*racmd();
static	int 	rdinit;
static	long	rdonline[8];

/*
 * XXX - Not consistent with other stand alone drivers.
 * static int	rd_boff[] = { 0, 9700, 27000, 30100, 0, 7460, 9700, 0 };
 */

raopen(io)
	register struct iob *io;
{
	register struct mscp *mp;
	int i;

	if (rdinit == 0) {
		RDADDR->raip = 0;
		while ((RDADDR->rasa & RA_STEP1) == 0)
			continue;
		RDADDR->rasa = RA_ERR;
		while ((RDADDR->rasa & RA_STEP2) == 0)
			continue;
		RDADDR->rasa = (short)&rd.ra_ca.ca_ringbase;
		while ((RDADDR->rasa & RA_STEP3) == 0)
			continue;
		RDADDR->rasa = (short)(segflag & 3);
		while ((RDADDR->rasa & RA_STEP4) == 0)
			continue;
		RDADDR->rasa = RA_GO;
		rd.ra_ca.ca_rspl = (short)&rd.ra_rsp.m_cmdref;
		rd.ra_ca.ca_rsph = (short)(segflag & 3);
		rd.ra_ca.ca_cmdl = (short)&rd.ra_cmd.m_cmdref;
		rd.ra_ca.ca_cmdh = (short)(segflag & 3);
		rd.ra_cmd.m_cntflgs = 0;
		if (racmd(M_O_STCON) < 0) {
			printf("RD: controller init error, STCON\n");
			return(-1);
		}
		rdinit = 1;
	}

	if (rdonline[io->i_unit & 7] == 0)
		if (ramount(io) == -1)
			return(-1);

	/*
	 * XXX - Not consistent with other stand alone drivers.
	 * if ((io->i_boff < 0) || (io->i_boff > 7) ||
	 *    (rd_boff[io->i_boff] == -1)) {
	 *	printf("RD: bad partition for unit=%d", io->i_unit & 7);
	 * }
	 * io->i_boff = rd_boff[io->i_boff];
	 */

	return(0);
}

raclose(io)
	register struct iob *io;
{
	rdonline[io->i_unit & 7] = 0;
	return(0);
}

ramount(io)
	register struct iob *io;
{
	if (racmd(M_O_ONLIN, io->i_unit) < 0) {
		printf("RD: bring online error, unit=%d\n", io->i_unit & 7);
		return(-1);
	}
	rdonline[io->i_unit & 7] = rd.ra_rsp.m_uslow +  
		((long)(rd.ra_rsp.m_ushigh) << 16);
	return(0);
}

struct mscp *
racmd(op, unit)
	int op, unit;
{
	register struct mscp *mp;
	register int i;

	rd.ra_cmd.m_opcode = op;
	rd.ra_cmd.m_unit = unit & 7;
	rd.ra_rsp.m_header.ra_msglen = sizeof(struct mscp);
	rd.ra_cmd.m_header.ra_msglen = sizeof(struct mscp);
	rd.ra_ca.ca_rsph = RA_OWN | RA_INT | (segflag & 3);
	rd.ra_ca.ca_cmdh = RA_OWN | RA_INT | (segflag & 3);
	i = RDADDR->raip;
	while (1) {
		if (rd.ra_ca.ca_cmdint)
			rd.ra_ca.ca_cmdint = 0;
		if (rd.ra_ca.ca_rspint)
			break;
	}
	rd.ra_ca.ca_rspint = 0;
	mp = &rd.ra_rsp;
	if (((mp->m_opcode & 0xff) != (op | M_O_END)) ||
	    ((mp->m_status & M_S_MASK) != M_S_SUCC)) {
		printf("RD: command error, unit=%d, opcode=%x, status=%x\n",
			unit & 7, mp->m_opcode & 0xff, mp->m_status);
		return((struct mscp *)-1);
	}
	return(0);
}

rastrategy(io, func)
	register struct iob *io;
{
	register struct mscp *mp;

	if (io->i_bn >= rdonline[io->i_unit & 7])
		return(0);

	mp = &rd.ra_cmd;
	mp->m_lbn_l = loint(io->i_bn);
	mp->m_lbn_h = hiint(io->i_bn);
	mp->m_bytecnt = io->i_cc;
	mp->m_buf_l = (ushort)io->i_ma;
	mp->m_buf_h = segflag & 3;
	if (racmd(func == READ ? M_O_READ : M_O_WRITE, io->i_unit) < 0)
		return(-1);

	return(io->i_cc);
}
